### PR TITLE
Add power simulation support and Verilator to the CI pipeline

### PR DESCRIPTION
**Pull request by ivysochyn**
# Add power simulation support and Verilator to the CI pipeline

This PR enhances the test suite by introducing `simulate` and `power` build targets for designs that include a `power.tcl` file. The `power.tcl` file sets the pass/fail criteria for power simulations, ensuring that tests are run only for designs with these predefined criteria.

This PR modifies the following:

* `flow/test/test_helper.sh` - Updated to execute power simulation for designs that have a `power.tcl` file present
* `docker/Dockerfile.builder` - Adds build and installation instructions for the Verilator
* `etc/DependencyInstaller.sh` - Updates dependencies to cover those of Verilator
* `.gitmodules` - Adds Verilator as a sub-module located under the `tools/verilator` path

### CODE REVIEW

**Review by oharboe — CHANGES_REQUESTED** [review]
Why can't this be part of ./setup.sh and the docker image, just like other tools provided by the OS?

Why should it be a submodule?

There are LOTS of dependencies in ORFS that are not submodules.

File: flow/test/test_helper.sh [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 set -eoux pipefail
 
-cd "$(dirname $(readlink -f $0))/../"
+cd "$(dirname "$(readlink -f "$0")")/../"
```
Comment (oharboe) [inline flow/test/test_helper.sh:6]: split into two commits: one for fixes, one for features, for simpler review.

File: flow/test/test_helper.sh [hunk 2]
```diff
@@ -59,8 +60,13 @@ fi
...
 
-if [ ! -z ${MAKE_ISSUE+x} ]; then
-  $__make final_report_issue 2>&1 | tee -a $LOG_FILE
+if [ -n "${MAKE_ISSUE+x}" ]; then
+  $__make final_report_issue 2>&1 | tee -a "$LOG_FILE"
+fi
+
+if [ $ret -eq 0 ] && [ -f "$POWER_TCL_FILE" ]; then
+  $__make simulate power 2>&1 | tee -a "$LOG_FILE"
+  ret=$?
```
Comment (vvbandeira) [inline flow/test/test_helper.sh]:
```suggestion
  ret=$(( ret + $? ))
```

**Review by vvbandeira — CHANGES_REQUESTED** [review]
Avoid overwriting previous return values.

File: etc/DependencyInstaller.sh [hunk 3]
```diff
@@ -34,6 +35,23 @@ _installCommon() {
     else
         pip3 install --no-cache-dir --user -U $pkgs
     fi
+
+    baseDir=$(mktemp -d /tmp/DependencyInstaller-orfs-XXXXXX)
+
+    # Install Verilator
```
Comment (oharboe) [inline etc/DependencyInstaller.sh:41]: Nice. I think that we're asking for less trouble by building ourselves than to use the version that comes with the OS distribution.

**Review by oharboe — APPROVED** [review]
Nit, better to split commit into a fixes and a feature commit to simplify review.

File: etc/DependencyInstaller.sh [hunk 4]
```diff
@@ -34,6 +35,23 @@ _installCommon() {
...
+    verilogPrefix=`realpath ${PREFIX:-"/usr/local"}`
```
Comment (maliberty) [inline etc/DependencyInstaller.sh]: `verilatorPrefix` makes more sense to me

File: etc/DependencyInstaller.sh [hunk 5]
```diff
@@ -34,6 +35,23 @@ _installCommon() {
     else
         pip3 install --no-cache-dir --user -U $pkgs
     fi
+
+    baseDir=$(mktemp -d /tmp/DependencyInstaller-orfs-XXXXXX)
```
Comment (maliberty) [inline etc/DependencyInstaller.sh:39]: Why is this needed?  It is already set in the script elsewhere.

File: etc/DependencyInstaller.sh [hunk 6]
```diff
@@ -71,19 +89,28 @@ _installUbuntuPackages() {
     export DEBIAN_FRONTEND="noninteractive"
     apt-get -y update
     apt-get -y install --no-install-recommends \
+        bison \
         curl \
+        flex \
+        help2man \
+        libfl-dev \
+        libfl2 \
+        libgoogle-perftools-dev \
```
Comment (maliberty) [inline etc/DependencyInstaller.sh:98]: Does Verilator require this?

File: etc/DependencyInstaller.sh [hunk 7]
```diff
@@ -215,6 +242,7 @@ while [ "$#" -gt 0 ]; do
             ;;
         -prefix=*)
             OR_INSTALLER_ARGS="${OR_INSTALLER_ARGS} $1"
+            PREFIX=${1#*=}
```
Comment (maliberty) [inline etc/DependencyInstaller.sh:245]: What does this do?

File: flow/test/test_helper.sh [hunk 8]
```diff
@@ -59,8 +60,13 @@ fi
...
 set -e
 
-if [ ! -z ${MAKE_ISSUE+x} ]; then
-  $__make final_report_issue 2>&1 | tee -a $LOG_FILE
+if [ -n "${MAKE_ISSUE+x}" ]; then
+  $__make final_report_issue 2>&1 | tee -a "$LOG_FILE"
+fi
+
+if [ $ret -eq 0 ] && [ -f "$POWER_TCL_FILE" ]; then
+  $__make simulate power 2>&1 | tee -a "$LOG_FILE"
```
Comment (maliberty) [inline flow/test/test_helper.sh]: Where is this target defined?

**Review by maliberty — CHANGES_REQUESTED** [review]

File: flow/test/test_helper.sh (hunk 8 above)
Comment (eszpotanski) [reply flow/test/test_helper.sh]: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/designs/asap7/mock-array/config.mk#L50

File: etc/DependencyInstaller.sh (hunk 6 above)
Comment (eszpotanski) [reply etc/DependencyInstaller.sh:98]: It's optional, but "should be installed for good performance" ([Verilator install prerequisites](https://verilator.org/guide/latest/install.html#install-prerequisites))

**Comment by maliberty** [comment]
Is there a reason to make this narrowly about a single design rather than a framework any design could use?
What is the runtime overhead of using Verilator?  Why not simply capture a vcd file?

**Comment by oharboe** [comment]
> Is there a reason to make this narrowly about a single design rather than a framework any design could use? What is the runtime overhead of using Verilator? Why not simply capture a vcd file?

- Any design can add simulation if they want under this framework.
- A .vcd file has to be regenerated via simulation after every time the ORFS flow has run. The test bench that exercises the design is design specific.
- What do you mean "overhead" of Verilator? Something has to run the simulation, Verilator is doing that. Are you asking how long this simulation stage takes in addition to the normal tests?

File: flow/test/test_helper.sh [hunk 9]
```diff
@@ -59,8 +60,13 @@ fi
 # Only enabled abort on error at the end to allow script to reach make issue
 set -e
 
-if [ ! -z ${MAKE_ISSUE+x} ]; then
-  $__make final_report_issue 2>&1 | tee -a $LOG_FILE
+if [ -n "${MAKE_ISSUE+x}" ]; then
+  $__make final_report_issue 2>&1 | tee -a "$LOG_FILE"
+fi
+
+if [ $ret -eq 0 ] && [ -f "$POWER_TCL_FILE" ]; then
```
Comment (oharboe) [inline flow/test/test_helper.sh]: Here it would be preferable to check if the make file with the current design has a "simulate" and "power" target and let the design implement these targets however it wants.

**Review by oharboe — COMMENTED** [review]
Suggestion to check for existance of "power" and "simulate" targets per design.

File: flow/test/test_helper.sh (hunk 1 above)
Comment (eszpotanski) [reply flow/test/test_helper.sh:6]: Commit has been split

File: flow/test/test_helper.sh (hunk 2 above)
Comment (eszpotanski) [reply flow/test/test_helper.sh]: Suggestion applied

File: etc/DependencyInstaller.sh (hunk 4 above)
Comment (eszpotanski) [reply etc/DependencyInstaller.sh]: changed to `verilatorPrefix`

File: flow/test/test_helper.sh [hunk 10]
```diff
@@ -12,6 +12,7 @@ if [ $# -eq 4 ]; then
   FLOW_VARIANT=$4
 fi
 DESIGN_CONFIG=./designs/$PLATFORM/$DESIGN_NAME/$CONFIG_MK
+POWER_TCL_FILE=./designs/$PLATFORM/$DESIGN_NAME/power.tcl
```
Comment (oharboe) [inline flow/test/test_helper.sh]: Deleted line above, unused.

File: etc/DependencyInstaller.sh (hunk 5 above)
Comment (eszpotanski) [reply etc/DependencyInstaller.sh:39]: I believe that `DependencyInstaller` from OpenROAD has this variable set once, but in ORFS each function has its own definitions of `baseDir`

**Review by oharboe — CHANGES_REQUESTED** [review]
nit

File: etc/DependencyInstaller.sh (hunk 7 above)
Comment (eszpotanski) [reply etc/DependencyInstaller.sh:245]: it assigns everything after `=` to the `PREFIX` variable, which is later used as a Verilator prefix installation

File: flow/test/test_helper.sh (hunk 9 above)
Comment (eszpotanski) [reply flow/test/test_helper.sh]: changed to separately checking whether make target exist

File: flow/test/test_helper.sh (hunk 10 above)
Comment (eszpotanski) [reply flow/test/test_helper.sh]: removed

**Comment by oharboe** [comment]
On my laptop:

- 500s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk `
- 256s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk simulate`
- 17s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk power`

Simulate can be sped up to 60s by this change, but I'd want to ask @vvbandeira whether that is a good idea... If there are lots of other processes on that CI server, then better to run in serial? It would improve the local user experience though.

```
$ git diff
diff --git a/flow/designs/asap7/mock-array/simulate.sh b/flow/designs/asap7/mock-array/simulate.sh
index 0796d6ec..8778b89d 100755
--- a/flow/designs/asap7/mock-array/simulate.sh
+++ b/flow/designs/asap7/mock-array/simulate.sh
@@ -27,6 +27,6 @@ verilator -Wall --cc \
   --exe \
   ../../../designs/src/mock-array/simulate.cpp
 
-make -C obj_dir -f VMockArray.mk
+make -j -C obj_dir -f VMockArray.mk
 
 obj_dir/VMockArray
```

**Comment by vvbandeira** [comment]
> On my laptop:
> 
> * 500s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk `
> * 256s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk simulate`
> * 17s `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk power`
> 
> Simulate can be sped up to 60s by this change, but I'd want to ask @vvbandeira whether that is a good idea... If there are lots of other processes on that CI server, then better to run in serial? It would improve the local user experience though.
> 
> ```
> $ git diff
> diff --git a/flow/designs/asap7/mock-array/simulate.sh b/flow/designs/asap7/mock-array/simulate.sh
> index 0796d6ec..8778b89d 100755
> --- a/flow/designs/asap7/mock-array/simulate.sh
> +++ b/flow/designs/asap7/mock-array/simulate.sh
> @@ -27,6 +27,6 @@ verilator -Wall --cc \
>    --exe \
>    ../../../designs/src/mock-array/simulate.cpp
>  
> -make -C obj_dir -f VMockArray.mk
> +make -j -C obj_dir -f VMockArray.mk
>  
>  obj_dir/VMockArray
> ```

Each design runs on its own VM, so it should be fine to run in parallel as long as they do not consume resources to the point of starving themselves.

**Comment by oharboe** [comment]
@maliberty mock-array finishes before other slower designs, so I don't see a reason to complicate this PR by speeding up `make simulate`

image

**Comment by maliberty** [comment]
I'm unclear why we are putting this in flow/test/test_helper.sh and not in the flow proper.  Users don't use test_helper.sh, they use make or bazel to run flow steps.

**Comment by oharboe** [comment]
> I'm unclear why we are putting this in flow/test/test_helper.sh and not in the flow proper. Users don't use test_helper.sh, they use make or bazel to run flow steps.

What is "in the flow proper"?

The idea of putting it in test_helper.sh is simply to get CI to pick it up.

Users would run "make simulate power". The reason config.mk of the design implements "simulate" and "power" is that it is a go-no go test where power is optional after simulate and it is highly design specific, so I'm unsure what code can be shared, at least until I see a couple of designs doing the same thing.

I think it makes sense to start with one design under CI for power, then add two three more over time to see what is generic and what is specific to the design. If we try to design something generic without having a few examples, we're sure to have to throw away our overcomplicated general design once we have a few examples as I'm sure they won't fit(because we design a general system with no examples, which is an exercise in futility...).

Can you suggest an alternative way to do this or is this a good 'nuf start?

We have:

- "make power simulate" under CI, if they exist
- we allow users to run the tests locally

**Comment by maliberty** [comment]
> What is "in the flow proper"?

As it stands we will do report_power and analyze_power_grid in final_report.tcl without any vcd.  Then we will do report_power again with a vcd only if the user does some extra manual steps.  We will never do analyze_power_grid with the vcd.  It seems better to handle this in final_report.tcl.

**Comment by oharboe** [comment]
I see. Can you take delivery of this and rework it, we dont have the domain knowledge to integrate....

**Comment by maliberty** [comment]
My comments are to be addressed in a follow on PR
